// react-docgen-typescript-loader/dist/validateOptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ajv_1 = __importDefault(require("ajv"));
var ajv = new ajv_1.default();
var validate = ajv.compile({
    type: "object",
    additionalProperties: false,
    properties: {
        skipPropsWithName: {
            anyOf: [
                { type: "string", minLength: 1 },
                { type: "array", minItems: 1, items: { type: "string", minLength: 1 } },
            ],
        },
        skipPropsWithoutDoc: {
            type: "boolean",
        },
        tsconfigPath: {
            type: "string",
            minLength: 1,
        },
        compilerOptions: {
            type: "object",
        },
        docgenCollectionName: {
            anyOf: [{ type: "string", minLength: 1 }, { type: "null" }],
        },
        setDisplayName: {
            type: "boolean",
        },
        includes: {
            type: "array",
            items: {
                type: "string",
            },
        },
        excludes: {
            type: "array",
            items: {
                type: "string",
            },
        },
    },
});
function validateOptions(options) {
    if (!options)
        return;
    var valid = validate(options);
    if (!valid) {
        var errorMessage = validate
            .errors.map(function (e) {
            var message = e.message;
            if (e.dataPath)
                message = e.dataPath + " " + message;
            return message;
        })
            .join(", ");
        throw new Error(errorMessage);
    }
}
exports.default = validateOptions;
